File: packages/ios-client/Sources/FishjamClient/media/Tracks/LocalVideoTrack.swift
import WebRTC

/// Utility wrapper around a local `RTCVideoTrack` also managing an instance of `VideoCapturer`
public class LocalVideoTrack: VideoTrack, LocalTrack {
    internal var capturer: CameraCapturer
    internal var videoParameters: VideoParameters
    internal var videoSource: RTCVideoSource

    internal init(
        mediaTrack: RTCVideoTrack, videoSource: RTCVideoSource, endpointId: String, metadata: Metadata = Metadata(),
        videoParameters: VideoParameters, capturer: CameraCapturer

    ) {
        self.videoParameters = videoParameters
        self.capturer = capturer
        self.videoSource = videoSource
        super.init(mediaTrack: mediaTrack, endpointId: endpointId, rtcEngineId: nil, metadata: metadata)
    }

    convenience internal init(mediaTrack: RTCVideoTrack, oldTrack: LocalVideoTrack) {
        self.init(
            mediaTrack: mediaTrack, videoSource: oldTrack.videoSource, endpointId: oldTrack.endpointId,
            metadata: oldTrack.metadata, videoParameters: oldTrack.videoParameters, capturer: oldTrack.capturer)
    }

    internal var mirrorVideo: (_ shouldMirror: Bool) -> Void = { _ in } {
        didSet {
            capturer.mirrorVideo = mirrorVideo
        }
    }

    public var isFrontCamera: Bool {
        return capturer.isFront
    }

    public func start() {
        capturer.startCapture()
    }

    public func stop() {
        capturer.stopCapture()
    }

    public func flipCamera() {
        capturer.switchCamera()
    }

    public func switchCamera(deviceId: String) {
        capturer.switchCamera(deviceId: deviceId)
    }

    public static func getCaptureDevices() -> [AVCaptureDevice] {
        return RTCCameraVideoCapturer.captureDevices()
    }
}

File: packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/media/LocalVideoTrack.kt
package com.fishjamcloud.client.media

import android.content.Context
import com.fishjamcloud.client.models.Metadata
import com.fishjamcloud.client.models.VideoParameters
import com.fishjamcloud.client.utils.getEnumerator
import org.webrtc.VideoSource

class LocalVideoTrack(
  mediaTrack: org.webrtc.VideoTrack,
  endpointId: String,
  metadata: Metadata,
  private val capturer: CameraCapturer, // TODO: This need to be refactor and changed to VideoCapturer
  val videoParameters: VideoParameters
) : VideoTrack(mediaTrack, endpointId, null, metadata),
  LocalTrack {
  val videoSource: VideoSource
    get() = capturer.source

  var captureDeviceChangedListener: CaptureDeviceChangedListener?
    get() = capturer.captureDeviceChangedListener
    set(value) {
      capturer.captureDeviceChangedListener = value
    }

  constructor(mediaTrack: org.webrtc.VideoTrack, oldTrack: LocalVideoTrack) : this(
    mediaTrack,
    oldTrack.endpointId,
    oldTrack.metadata,
    oldTrack.capturer,
    oldTrack.videoParameters
  )

  companion object {
    fun getCaptureDevices(context: Context): List<CaptureDevice> {
      val enumerator = getEnumerator(context)
      return enumerator.deviceNames.map { name ->
        CaptureDevice(
          name,
          enumerator.isFrontFacing(name),
          enumerator.isBackFacing(name)
        )
      }
    }
  }

  override fun start() {
    capturer.startCapture()
  }

  override fun stop() {
    capturer.stopCapture()
  }

  suspend fun flipCamera() {
    capturer.flipCamera()
  }

  suspend fun switchCamera(deviceName: String) {
    capturer.switchCamera(deviceName)
  }

  fun isFrontCamera(): Boolean = capturer.isFrontFacingCamera ?: false

  fun getCaptureDevice(): CaptureDevice? = capturer.getCaptureDevice()
}
